File: C++/AtCoder/ふるい/DPCM_1117.cpp
// copyright
#include <bits/stdc++.h>

int main() {
    std::vector<int64_t> ans_list;
    int y_max = 255;
    int y_min = 0;
    int64_t INF = 1e18;

    while (true) {
        int n;
        int m;
        std::cin >> n >> m;
        if (n == 0 && m == 0) {
            break;
        }

        int C[m];
        int X[n];
        for (int i = 0; i < m; i++) {
            std::cin >> C[i];
        }
        for (int i = 0; i < n; i++) {
            std::cin >> X[i];
        }

        int64_t DP[n+1][y_max+1];
        for (int i = 0; i <= n; i++) {
            for (int j = 0; j <= y_max; j++) {
                DP[i][j] = INF;
            }
        }

        DP[0][128] = 0;

        for (int pre_ind = 0; pre_ind < n; pre_ind++) {
            for (int pre_y = 0; pre_y <= y_max; pre_y++) {
                int nex_ind = pre_ind + 1;
                for (auto c : C) {
                    int nex_y;
                    nex_y = pre_y + c;
                    nex_y = std::max(y_min, std::min(y_max, nex_y));
                    DP[nex_ind][nex_y] = std::min(DP[nex_ind][nex_y], DP[pre_ind][pre_y] + (X[nex_ind-1] - nex_y) * (X[nex_ind-1] - nex_y));
                }
            }
        }

        int64_t ans = INF;
        for (int64_t dp : DP[n]) {
            ans = std::min(ans, dp);
        }
        ans_list.push_back(ans);
    }

    for (auto ans : ans_list) {
        std::cout << ans << std::endl;
    }
}

File: Python/AtCoder/old/DPCM_1117.py
# y: 複合化後の音声信号
# C: コードブック
# k:出力系列 今回選ぶもの
# y[n] = y[n-1]+C[k[n]]
# y[0] = 128
y_max = 255
y_min = 0
INF = 10**9


def main():
    ans_list = []
    while True:
        n, m = map(int, input().split())
        if (n, m) == (0, 0):
            break
        C = [int(input()) for _ in range(m)]
        X = [int(input()) for _ in range(n)]

        # DP[ind][y] = X[ind]まで見たときで、末尾がyとなるときの二乗和の最小値
        # DP[ind][y] = DP[ind-1][y-C[x]] + (X[ind] - y)**2
        DP = [[INF]*(y_max+1) for _ in range(n+1)]

        # 初期化
        DP[0][128] = 0

        for pre_ind in range(n):
            for pre_y in range(y_max+1):
                nex_ind = pre_ind + 1
                for c in C:
                    nex_y = pre_y + c
                    nex_y = max(y_min, min(y_max, nex_y))
                    DP[nex_ind][nex_y] = min(DP[nex_ind][nex_y], DP[pre_ind][pre_y] + (X[nex_ind-1] - nex_y)**2)

        ans = min(DP[n])
        ans_list.append(ans)
        # print(ans)

    print(*ans_list, sep='\n')


main()
